cpp/include/raft/core/device_container_policy.hpp

```diff
@@ -164,10 +164,19 @@ class device_uvector_policy {
  public:
   auto create(raft::resources const& res, size_t n) -> container_type
   {
-    return container_type(n, resource::get_cuda_stream(res), resource::get_workspace_resource(res));
+    if (mr_ == nullptr) {
```

Comment:
You're right, this extra field in container policy increases the footprint of all device mdarrays.

The main reason for having it as a member is to not change the mdarray api. This is, by the way, also the way we used to have it before the introduction of the workspace memory resources.

I think you raise a very good point here, and I would be up to do the necessary change. I think, the minimum required change is to add an extra constructor to mdarray that would take the memory resource. However, there are couple problems with this:
  - The dependent types, such as raft sparse matrices, would need to be modified as well to allow them to specify the memory resource.
  - This also breaks the encapsulation: normally an mdarray should rely only on the container policy type for managing allocations, but with the change, there would be an explicit memory-related argument passed to mdarray besides the container policy.
@cjnolet , what are your thoughts on this?

Reply:
I don't think we want to change the mdarray API, especially since we're eventually planning to consolidate w/ CCCL's mdarray (which will already be painful, but the more we diverge from the reference/standard, the more painful it will become). 

One thing we can do is maybe split the current container policy into two strategies- the standard device_containter_policy that just uses the `get_default_resource` and perhaps one that explicitly uses the `get_workspace_resource`? What do you think? I think it's important that we couple the container policy to the `raft::resources` as much as possible so that we limit the amount of potential places where different state is initialized- if we need a container policy with a different memory resource, it should come from the `raft::resources` layer.

Reply:
The problem is that the container policy is a template parameter of mdarray and stored by value inside it (we cannot just inherit it). In raft we have many helper functions and wrapper types, which fix the type of the container policy. If we're to make several different policies now, that would be a major overhaul. Thus I suggest we keep it as it is for now, and discuss whether we want to do this in a follow-up PR.

Reply:
Just to clarify my stance here- I'm not suggesting inheritance here- the relationship could be from a composition. I'm just saying that it might make more sense to provide different container policies for workspace and default memory resource variants. We could also provide new helpers for constructing the "workspace" variants if that would make it easier. That being said, I'm okay with pushing that one off until we have had more time to think about and discuss it.

Reply:
Independently of how we do it, the type of the policy is the part of the mdspan template. We have all those synonyms for device_mdspans and host_mdspans that hide this and fix the type of the policy. We would need to change all of those if we want to allow more than one policy.
Thanks, let's push this off for the next release.

Reply:
Thanks for the discussion. Agreed, we can proceed for now with the current implementation.